### ISSUE TITLE
Lock screen timeout:  Option to Sign Out

### ISSUE THREAD

**Issue by joelbrock** [CORE-POS/IS4C#1020]
In Extra Configs we can configure a timeout after which the register screen LOCKs itself.

Several have requested the option to configure this timeout to trigger a SIGN OUT, as opposed to a LOCK.

I imagine it could be one addt'l checkbox on the extra_config.php page.  And we could use the timeout value (e.g. 180000) as double-duty for both lock and sign out.
Obviously only one option should work at a time.  So maybe it’s a radio button?

**Comment by gohanman** [comment]
What's the desired behavior if the timeout occurs mid-transaction? Bringing the next user to a fresh transaction and having to re-ring things doesn't seem ideal. The easier solution might be to revert to just locking if items are present; Otherwise the next user could take over the existing transaction and update the already rung items with their employee # etc but that's going to touch on more stuff.

**Comment by joelbrock** [comment]
yeah that sounds right:  LOCK if there are items present.  SIGN OUT if no items.

**Comment by joshuabay** [comment]
I created a working draft of this at our co-op and am curious if you have any input. 

In [pos2.js#L65](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/gui-modules/js/pos2.js#L65), I edited the lockscreen function to check if items are present by polling `LastID`. 

If items are present, it will LOCK. If not, SIGN OUT. 

```javascript
    var lockScreen = function() {
        // poll LastID to determine if items are present
        var url = urlStem + 'ajax/AjaxSession.php?var=LastID';
        $.get( url , function( data ) {
            var lastID = data;
            console.log( "lastID = " + lastID );
            // if number of items is greater than 0, lock screen, otherwise sign out
            if (lastID > 0) {
            window.location = urlStem + 'gui-modules/login3.php';
            } else {
            window.location = urlStem + 'gui-modules/pos2.php?reginput=SO&repeat=1';
            }
        });
    };
```

I could not find an existing Ajax Callback to retrieve the `LastID` Session Variable, so I created a simple one:

```php
// AjaxSession.php
<?php
session_start();
$var = htmlentities($_GET['var']);
echo $_SESSION[$var];
```

I realize this is not up to coding standards and should probably be rewritten as a plugin. 

How should this be reworked if I wanted to eventually submit it to the public repository?

**Comment by gohanman** [comment]
It looks pretty good to me. There's not really a good way to do this as a plugin because there's not a way to hook into that lock screen function.

I'd do the ajax side more like this:
```php
namespace COREPOS\pos\ajax;
use COREPOS\pos\lib\AjaxCallback;
include_once(dirname(__FILE__).'/../lib/AutoLoader.php');

class AjaxSession extends AjaxCallback
{ 
    protected $encoding = 'plain'; 

    private $allowedKeys = array(
        'LastID' => true,
    );

    public function ajax()
    {
        $input = $this->form->tryGet('var');
        if (isset($this->allowedKeys($input))) {
            return $this->session->get($input);
        }

        return '';
    }
}

AjaxSession::run();
```

Most of the structuring is just to make it include-safe. And keeping some restriction on what value the client side is allowed to pull from the session seems like a good idea.

I think the only other thing it would need is a setting to turn the new behavior on and off. Since the setting would be on the PHP side you'd probably need:
```javascript
var pos2 = (function($, errorLog, lockBehavior) {
```
in place of 
```javascript
var pos2 = (function($, errorLog) {
```
and then adjust the call to it in `pos2.php` to pass in the new setting.

**Comment by joshuabay** [comment]
Thanks for the input. I'll take your approach to AjaxSession and add the lockBehavior setting to turn it on/off.
